File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    public int ROWS = 0;
        public int COLS = 0;
        public int fresh = 0;
        public int[][] directions = new int[][]{{-1,0},{0,1},{1,0},{0,-1}};
        public int[][] gridcpy;
        public Queue<int[]> queue = new LinkedList<>();
        public int returnValue = 0;
    public Solution(){
        
        
    }
    public int orangesRotting(int[][] grid) {
        //1. Initialize the grid values 
        this.ROWS = grid.length;
        this.COLS = grid[0].length;
        this.gridcpy = grid;
        //2. For loop 
        for(int r = 0;r<this.ROWS;r++){
            for(int c = 0;c<this.COLS;c++){
                if(grid[r][c] == 2){
                    this.queue.offer(new int[]{r,c});
                }
                else if(grid[r][c]==1){
                    this.fresh++;
                }
            }
        }
        
        //3. Call BFS function 
        this.bfs();
    
        return this.fresh==0?this.returnValue:-1;
        
    }
    
    public void bfs(){
        //1. Initialize values in q = DONE
        //2. 9While queue
        while((!this.queue.isEmpty()) && (this.fresh>0)){
            //loop for the level 
            this.returnValue++;
            int size = this.queue.size();
            for(int i=0;i<size;i++){
                //1. Pop the first element 
                int[] current = this.queue.poll();
                //2. loop throguh the neighbors
                for(int[] arr: this.directions){
                    int newR = current[0] + arr[0];
                    int newC = current[1] + arr[1];
                    
                    //3. check the boundaries 
                    if((newR>=0)&&(newR<this.ROWS)&&(newC>=0)&&(newC<this.COLS)){
                        if(this.gridcpy[newR][newC] == 1){
                            this.gridcpy[newR][newC] = 2;
                            this.queue.offer(new int[]{newR,newC});
                            this.fresh--;
                        }
                    }
                }
            }
        }
    }
    
}

File: 0994-rotting-oranges/0994-rotting-oranges.py
from collections import deque
class Solution(object):
    def __init__(self):
        self.ROWS = 0
        self.COLS = 0
        self.queue = deque()
        self.fresh = 0
        self.minutes = 0
        self.gridcpy = []
        self.directions = [(-1,0),(1,0),(0,-1),(0,1)]
        
        
    def orangesRotting(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        self.gridcpy = grid
        self.ROWS = len(grid)
        self.COLS = len(grid[0])
        #1. Iterate through all the cells in the grid 
        for r in range(self.ROWS):
            for c in range(self.COLS):
                if grid[r][c] == 1:
                    self.fresh+=1 
                elif grid[r][c] == 2:
                    self.queue.append((r,c))
        
        self.bfs()
        
        #2. Return the minutes 
        return self.minutes if self.fresh==0 else -1
    
    def bfs(self):
        #1. Initialize things in the queue 
        # DONE
        
        #2. Whle queue 
        while self.queue and self.fresh>0:
            # Increment the time at start of each level and loop through all the levels
            self.minutes+=1 
            for _ in range(len(self.queue)):
            
                cRow, cCol = self.queue.popleft()

                # loop through all the directions
                for direction in self.directions:
                    iRow, iCol = direction
                    row = cRow + iRow
                    col = cCol + iCol

                    if 0<=row<self.ROWS and 0<=col<self.COLS:
                        if self.gridcpy[row][col] == 1:
                            self.fresh-=1
                            self.gridcpy[row][col] = 2
                            self.queue.append((row, col))
        
    
                    
        